torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -254,34 +277,74 @@ def _post_backward(self, *unused: Any):
             )
 
     def finalize_backward(self):
-        if self._sharded_state == ShardedState.UNSHARDED:
-            # Run post-backward here since the forward inputs did not require
-            # gradient, so the post-backward hook did not run
-            self._post_backward()
         if self._reduce_scatter_view_out_event is not None:
             torch.cuda.current_stream().wait_event(self._reduce_scatter_view_out_event)
             self._reduce_scatter_view_out_event = None
         self._training_state = TrainingState.IDLE
+        self._post_forward_indices.clear()
+        self.all_forward_grad_fns.clear()
+
+    def _prefetch_unshard(self):
+        if self._training_state == TrainingState.PRE_BACKWARD:
+            if not self._post_forward_indices:
+                # Can be cleared if running multiple `backward`s
+                return
+            curr_index = self._post_forward_indices.pop()
+            if (target_index := curr_index - 1) < 0:
+                return
+            target_fsdp_param_group = self.comm_ctx.post_forward_order[target_index]
+            if any(
+                torch._C._will_engine_execute_node(grad_fn)  # type: ignore[attr-defined]
+                for forward_grad_fns in target_fsdp_param_group.all_forward_grad_fns
+                for grad_fn in forward_grad_fns
+            ):
+                with torch.profiler.record_function(
+                    "FSDP::backward_prefetch"
+                ), target_fsdp_param_group.use_training_state(
+                    TrainingState.PRE_BACKWARD
+                ):
+                    target_fsdp_param_group.unshard()
```

Comment:
hmmm I wonder if `target_fsdp_param_group` is always `self`? (I guess not). If it's not always self, it feels a bit weird to me that the current fsdp_param_group is actually prefetching another "target" fsdp_param_group

Reply:
I think it is never `self`. I wonder why it is weird haha -- the currently executing group has to prefetch on behalf of the next group since the next group did not run yet.

Reply:
Oh it's more like OO design issue, intuitively I feel prefetching should happen in a higher level abstraction that manages ParamGroups instead of ParamGroup directly, a ParamGroup instance should morally should only care about its own state (its own unshard/shard/reshard), instead of other param groups.

Reply:
Ah yes that makes sense. There is definitely some leaky abstraction here (as evidenced even by the circular references). I did not have a solution to reorganize though 🤔

Reply:
I think it's fine to land as is, this is something we could think about after the initial PRs landed